src/python/pants/binaries/binary_tool_mixin.py

```diff
@@ -0,0 +1,98 @@
+# coding=utf-8
+# Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from collections import namedtuple
+
+from pants.binaries.binary_util import BinaryUtil
+from pants.util.memo import memoized_method
+
+
+class BinaryToolMixin(object):
+  """Mixin for classes that use binary executables.
+
+  Must be mixed in to something that can register and use options, e.g., a Task or a Subsystem.
+  Specifically, the class this is mixed into must have a get_options() method.
+
+  :API: public
+  """
+  class BinaryTool(namedtuple('BinaryTool', ['scope', 'supportdir', 'name', 'platform_dependent',
+                                             'replaces_scope', 'replaces_name'])):
+
+    # TODO(benjy): Remove replaces_scope/replaces_name after migration to this mixin is complete.
+    def select(self, options):
+      version = getattr(options.for_scope(self.scope), '{}_version'.format(self.name))
+      if self.replaces_scope and self.replaces_name:
+        # If the old option is provided explicitly, let it take precedence.
+        old_opts = options.for_scope(self.replaces_scope)
+        if not old_opts.is_default(self.replaces_name):
+          version = old_opts.get(self.replaces_name)
+      return BinaryUtil.Factory.create().select(
+        self.supportdir, version, self.name, self.platform_dependent)
+
+  @staticmethod
+  def get_registered_tools():
+    """Returns a map of name to BinaryTool."""
+    return BinaryToolMixin._binary_tools
+
+  @staticmethod
+  def reset_registered_tools():
+    """Needed only for test isolation."""
+    BinaryToolMixin._binary_tools = {}
+
+  _binary_tools = {}  # name -> BinaryTool objects.
```

Comment:
This bit of global mutable state is unfortunate, because it will make it more challenging to move to v2. JVM bootstrap is maybe not the best thing to emulate in that regard.

It seems like there is a relatively clear path to avoiding putting mutable state in Tasks/Subsystems though: what about require tool registrations to be declarative at the `register.py` level? Or could we perhaps require this mixin to exist on a registered Optionable (ie, either a Task/Subsystem)?

If this were an operation outside of the task model, the only question would be how to trigger it... a bare invoke of `./pants --bootstrap-tools` would be a bit odd.

The downside of that would be that it would not be a native inhabitant of the "v1 task model" (or v2, for that matter), but it would seem that "static external binary tools" are a bit of a special case in terms of bootstrapping, regardless of implementation.

Reply:
I don't like the idea of doing this in register.py. That seems like a weird bit of out-of-band state to have to maintain. For example, if I write a base class that uses a binary tool then every author of every subclass has to know to do something special in register.py, because of an implementation detail of a base class that may change in the future. This really should be done in the Task code.

We can require this to be mixed into Task or Subsystem (and ~do, since it must be mixed in on something that has a get_options()). Not sure how that helps though? So at eager bootstrap time we iterate over all registered tasks and type-test them for the mixin? That seems feasible.

I can imagine a `./pants warmup` or something that does the eager bootstrapping.

Reply:
> So at eager bootstrap time we iterate over all registered tasks and type-test them for the mixin? That seems feasible.

Yea, that's basically what I was thinking.

It's just the mutable state that is problematic.

Reply:
We do need to do something with the map from logical name to `BinaryTool` instance somewhere in `register_options`, which is a classmethod. So this information has to live in the class, not on the task instance. But I think it's OK if it lives in the task class (rather than the `BinaryToolMixin` base class)? We can guarantee that this state is not mutated after `register_options` completes.

Reply:
@stuhood Have a look at the latest commit and see if that addresses your concerns?